File: lib/forward-auth.ts
import { NextRequest } from 'next/server';
import { prisma } from '@/lib/prisma';

export type ForwardAuthUser = {
  id: string;
  username: string;
  email: string | null;
  autheliaId: string;
  displayName: string;
  createdAt: Date;
  updatedAt: Date;
};

export async function getForwardAuthUser(request: NextRequest): Promise<ForwardAuthUser | null> {
  const autheliaId = request.headers.get('Remote-User');

  if (!autheliaId) {
    return null;
  }

  const username = autheliaId;
  const email = request.headers.get('Remote-Email');
  const displayName = request.headers.get('Remote-Name') ?? '';

  const existingUser = await prisma.user.findUnique({
    where: { autheliaId },
  });

  if (existingUser) {
    const updatedUser = await prisma.user.update({
      where: { autheliaId },
      data: { email, displayName },
    });
    return updatedUser;
  }

  // authentik → authelia 전환: 기존 사용자가 구 authentik UID로 등록되어 있을 수 있음
  // username으로 fallback 조회하여 autheliaId를 업데이트
  const existingByUsername = await prisma.user.findUnique({
    where: { username },
  });

  if (existingByUsername) {
    const updatedUser = await prisma.user.update({
      where: { username },
      data: { autheliaId, email, displayName },
    });
    return updatedUser;
  }

  const newUser = await prisma.user.create({
    data: { autheliaId, username, email, displayName },
  });

  return newUser;
}

File: backend/internal/auth/forward_auth.go
package auth

import (
	"context"
	"net/http"

	"github.com/dlddu/gatekeeper/backend/internal/httpx"
	"github.com/dlddu/gatekeeper/backend/internal/models"
	"github.com/dlddu/gatekeeper/backend/internal/store"
)

type ctxKey int

const userKey ctxKey = iota

// UserFromContext returns the authenticated user attached to the request context.
func UserFromContext(ctx context.Context) (*models.User, bool) {
	u, ok := ctx.Value(userKey).(*models.User)
	return u, ok
}

// ResolveForwardAuthUser looks up (or upserts) the user implied by the
// Authelia forward-auth headers. Returns nil when no Remote-User header is
// present, mirroring the behavior of lib/forward-auth.ts.
func ResolveForwardAuthUser(ctx context.Context, r *http.Request, users *store.UserStore) (*models.User, error) {
	autheliaID := r.Header.Get("Remote-User")
	if autheliaID == "" {
		return nil, nil
	}

	username := autheliaID
	displayName := r.Header.Get("Remote-Name")

	var emailPtr *string
	if v := r.Header.Get("Remote-Email"); v != "" {
		emailPtr = &v
	}

	if existing, err := users.FindByAutheliaID(ctx, autheliaID); err != nil {
		return nil, err
	} else if existing != nil {
		return users.UpdateProfileByAutheliaID(ctx, autheliaID, emailPtr, displayName)
	}

	// authentik → authelia migration: try matching by username.
	if existing, err := users.FindByUsername(ctx, username); err != nil {
		return nil, err
	} else if existing != nil {
		return users.UpgradeLegacyByUsername(ctx, username, autheliaID, emailPtr, displayName)
	}

	user := &models.User{
		AutheliaID:  autheliaID,
		Username:    username,
		Email:       emailPtr,
		DisplayName: displayName,
	}
	if err := users.Create(ctx, user); err != nil {
		return nil, err
	}
	return user, nil
}

// RequireForwardAuth is HTTP middleware that resolves the forward-auth user
// and attaches it to the request context. Unauthenticated requests get 401.
func RequireForwardAuth(users *store.UserStore, message string) func(http.Handler) http.Handler {
	if message == "" {
		message = "Unauthorized"
	}
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			user, err := ResolveForwardAuthUser(r.Context(), r, users)
			if err != nil {
				httpx.Error(w, http.StatusInternalServerError, "internal error")
				return
			}
			if user == nil {
				httpx.Error(w, http.StatusUnauthorized, message)
				return
			}
			ctx := context.WithValue(r.Context(), userKey, user)
			next.ServeHTTP(w, r.WithContext(ctx))
		})
	}
}
